lib/core/error/http/http-error.ts

```diff
@@ -0,0 +1,33 @@
+export class HttpError<TCode extends number = number> extends Error {
+  public readonly cause: unknown;
+  public readonly statusCode: TCode;
+  public readonly message: string;
+  public readonly url: string;
+  public readonly method: string;
+
+  constructor(opts: { url?: string; method?: string; message?: string; statusCode: TCode; cause?: unknown }) {
+    super(opts.message ?? `HTTP Error ${opts.statusCode} `);
+
+    Object.setPrototypeOf(this, HttpError.prototype);
+    this.name = HttpError.prototype.constructor.name;
+
+    this.cause = opts.cause;
+    this.statusCode = opts.statusCode;
+    this.url = opts.url;
+    this.method = opts.method;
+    this.message = opts.message;
+
+    if (opts.cause instanceof Error && opts.cause.stack) {
+      this.stack = opts.cause.stack;
+    }
+  }
+
+  static errorFactory(request, response: Response) {
```

Comment:
1. Type `request`
1. Name appropriately so that we are doing this from a request

```suggestion
  public static fromRequest(request: WhateverRequestIs, response: Response) {
```

Reply:
Yeah this is what I meant with the full FP approach, instead of throw new HttpError you  end up with `throw errorFactory(400)` - If you want to be fancy you can have Enums, but that isn't necessary in my opinion. This implementation excludes an error map (human readable strings) but it'll work.

Reply:
Yeah this is what I meant with the full FP approach, instead of throw new HttpError you end up with throw errorFactory(400) - If you want to be fancy you can have Enums, but that isn't necessary in my opinion. This implementation excludes an error map (human readable strings) but it'll work.